# Overall design	
# Single cell mRNA profiles of 6 CRPC patients were generated by 10X Genomics plartform, without technical replication, and sequenced by NovaSeq 6000 (Illumina)

setwd("GSE137829/")

rm(list = ls());gc()
set.seed(1011)
# load pkgs---------------------------------------------------------------------
library(dplyr)
library(Seurat)
library(Startrac)
require(Matrix)
library(SeuratWrappers)
library(clustree) # use for determine the optimal resolution
# library(ROGUE) # use for determine the optimal resolution
library(harmony)
library(stringr)
library(decontX)
library(scDblFinder)
library(DoubletFinder)
library(Augur)
library(DESeq2)
library(clusterProfiler)
library(enrichplot)
library(ggplot2)
library(ggpubr)
library(ggalluvial)
library(patchwork)
library(tidyr)
library(DESeq2)
require(stringr)
# load data---------------------------------------------------------------------
p1 <- data.table::fread("GSM4089151_P1_gene_cell_exprs_table.txt.gz")
p2 <- data.table::fread("GSM4089152_P2_gene_cell_exprs_table.txt.gz")
p3 <- data.table::fread("GSM4089153_P3_gene_cell_exprs_table.txt.gz")
p4 <- data.table::fread("GSM4089154_P4_gene_cell_exprs_table.txt.gz")
p5 <- data.table::fread("GSM4711414_P5_gene_cell_exprs_table.txt.gz")
p6 <- data.table::fread("GSM4711415_P6_gene_cell_exprs_table.txt.gz")

# lapply(list(p1, p2, p3, p4, p5, p6), dim)

colnames(p1)[3:ncol(p1)] <- paste0("p1-", colnames(p1)[3:ncol(p1)])
colnames(p2)[3:ncol(p2)] <- paste0("p2-", colnames(p2)[3:ncol(p2)])
colnames(p3)[3:ncol(p3)] <- paste0("p3-", colnames(p3)[3:ncol(p3)])
colnames(p4)[3:ncol(p4)] <- paste0("p4-", colnames(p4)[3:ncol(p4)])
colnames(p5)[3:ncol(p5)] <- paste0("p5-", colnames(p5)[3:ncol(p5)])
colnames(p6)[3:ncol(p6)] <- paste0("p6-", colnames(p6)[3:ncol(p6)])

rmdup <- function(data){
  data <- data[,-1]
  symbol.dup <- unique(data$Symbol[duplicated(data$Symbol)])
  data.dup <- data %>% filter(Symbol %in% symbol.dup)
  data.uni <- data %>% filter(!(Symbol %in% symbol.dup)) %>% 
    as.data.frame %>% 
    tibble::column_to_rownames("Symbol")
  
  data.dup.clean <- data.frame(
    row.names = colnames(data)[-1]
  ) %>% t %>% as.data.frame
  for(i in data$Symbol[duplicated(data$Symbol)]){
    tmp <- data.frame(
      expr = colSums(data[data$Symbol == i,2:ncol(data)]),
      row.names = colnames(data)[-1]
    ) %>% t %>% as.data.frame
    rownames(tmp) <- i
    
    print(all(colnames(tmp) == colnames(data.dup.clean)))
    
    data.dup.clean <- rbind(data.dup.clean, tmp)
  }
  
  if(all(colnames(data.uni) == colnames(data.dup.clean))){
    data.rmdup <- bind_rows(data.uni, data.dup.clean)
    return(data.rmdup) 
  }else{
    return(NULL)
  }
}

res <- lapply(list(p1 = p1, p2 = p2, p3 = p3, p4 = p4, p5 = p5, p6 = p6), rmdup)

p1 <- res$p1
p2 <- res$p2
p3 <- res$p3
p4 <- res$p4
p5 <- res$p5
p6 <- res$p6

symbols <- rownames(p1) %>%
  intersect(rownames(p2)) %>%
  intersect(rownames(p3)) %>%
  intersect(rownames(p4)) %>%
  intersect(rownames(p5)) %>%
  intersect(rownames(p6))

length(symbols)
# [1] 15709

p1$Symbol <- rownames(p1)
p2$Symbol <- rownames(p2)
p3$Symbol <- rownames(p3)
p4$Symbol <- rownames(p4)
p5$Symbol <- rownames(p5)
p6$Symbol <- rownames(p6)

dat <- p1 %>% filter(Symbol %in% symbols) %>% 
  full_join(p2 %>% filter(Symbol %in% symbols), by = "Symbol") %>% 
  full_join(p3 %>% filter(Symbol %in% symbols), by = "Symbol") %>% 
  full_join(p4 %>% filter(Symbol %in% symbols), by = "Symbol") %>% 
  full_join(p5 %>% filter(Symbol %in% symbols), by = "Symbol") %>% 
  full_join(p6 %>% filter(Symbol %in% symbols), by = "Symbol")

sum(duplicated(dat$Symbol))
# [1] 0

dim(dat)
# [1] 15709 27249

rownames(dat) <- dat$Symbol
dat <- dat %>% dplyr::select(-Symbol)

mtx <- dat %>% 
  as.matrix %>% 
  Matrix(sparse = TRUE)

metadata <- data.frame(
  orig.ident = str_remove_all(colnames(mtx), "-[ACTG]+.[0-9]+$"),
  row.names = colnames(mtx))
# Create Seurat object----------------------------------------------------------
scobj <- CreateSeuratObject(
  counts = mtx, 
  meta.data = metadata, 
  project = "GSE137829")

scobj
# An object of class Seurat 
# 15709 features across 27248 samples within 1 assay 
# Active assay: RNA (15709 features, 0 variable features)
# 1 layer present: counts

range(scobj$nFeature_RNA)
# [1]   49 9504

scobj.raw <- scobj
scobj[["RNA"]] <- split(scobj[["RNA"]], f = scobj$orig.ident)

range(scobj.raw$nFeature_RNA) # [1]   49 9504
range(scobj$nFeature_RNA) # [1]   49 9504

range(scobj.raw$nFeature_RNA[scobj.raw$orig.ident == "p1"]) # [1]   550 6152
range(scobj$nFeature_RNA[scobj$orig.ident == "p1"]) # [1]   550 6152


table(scobj$orig.ident)
# p1   p2   p3   p4   p5   p6 
# 3961 1397 1517 1641 9821 8911 
# QC----------------------------------------------------------------------------
scobj[["percent.mt"]] <- PercentageFeatureSet(scobj, pattern = "^MT-")
scobj[["percent.rp"]] <- PercentageFeatureSet(scobj, pattern = "^RP[SL]")
scobj[["percent.hb"]] <- PercentageFeatureSet(scobj, pattern = "^HB[^(P)]")

pdf("QC.pdf")
VlnPlot(scobj, group.by = "orig.ident", features = c("nFeature_RNA", "nCount_RNA","percent.mt","percent.rp","percent.hb"), ncol = 5)
VlnPlot(scobj, group.by = "orig.ident", features = c("nFeature_RNA"), ncol = 1)
VlnPlot(scobj, group.by = "orig.ident", features = c("nCount_RNA"), ncol = 1)
VlnPlot(scobj, group.by = "orig.ident", features = c("percent.mt"), ncol = 1) + scale_y_continuous(breaks = c(10,20))
VlnPlot(scobj, group.by = "orig.ident", features = c("percent.rp"), ncol = 1) + scale_y_continuous(breaks = c(10,20,30,40))
VlnPlot(scobj, group.by = "orig.ident", features = c("percent.hb"), ncol = 1)
FeatureScatter(scobj, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
dev.off()

scobj <- subset(
  scobj, 
  subset = 
    nFeature_RNA > 200 &
    nFeature_RNA < 7500 &
    nCount_RNA > 200 &
    nCount_RNA < 25000 &
    percent.mt < 10 &
    # percent.rp < 20 &
    percent.hb < 0.1)  

range(scobj$nFeature_RNA)
# [1]  206 5722
head(scobj@meta.data[,1:3])
# orig.ident nCount_RNA nFeature_RNA
# p1-AAACCTGAGAGCTATA.1         p1       8475         2237
# p1-AAACCTGAGTGTTTGC.1         p1      11489         2617
# p1-AAACCTGGTACATGTC.1         p1      13608         2531
# p1-AAACCTGGTCGGCATC.1         p1      15664         3196
# p1-AAACCTGGTGCAGACA.1         p1      13066         2851
# p1-AAACCTGTCATTCACT.1         p1      22751         4221

pdf("QC2.pdf")
VlnPlot(scobj, features = c("nFeature_RNA", "nCount_RNA","percent.mt","percent.rp","percent.hb"), ncol = 5)
VlnPlot(scobj, features = c("nFeature_RNA"), ncol = 1)
VlnPlot(scobj, features = c("nCount_RNA"), ncol = 1)
VlnPlot(scobj, features = c("percent.mt"), ncol = 1) + scale_y_continuous(breaks = c(10,20))
VlnPlot(scobj, features = c("percent.rp"), ncol = 1)
VlnPlot(scobj, features = c("percent.hb"), ncol = 1)
FeatureScatter(scobj, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
dev.off()

table(scobj$orig.ident)
# p1   p2   p3   p4   p5   p6 
# 3397  922 1319 1242 7904 6732

# dim(scobj)
# [1] 15709 21516
# blacklist---------------------------------------------------------------------
blacklist <- readxl::read_excel("C:/D/R project/Multi-Omics-Routine/scRNAseq/blacklist.xlsx")
sum(blacklist$Symbol %in% rownames(scobj))
# [1] 224

scobj <- scobj[!(rownames(scobj) %in% blacklist$Symbol),]
dim(scobj)
# [1] 15486 21516

range(scobj$nFeature_RNA)
# [1]  206 5722
head(scobj@meta.data[,1:3])
# orig.ident nCount_RNA nFeature_RNA
# p1-AAACCTGAGAGCTATA.1         p1       8475         2237
# p1-AAACCTGAGTGTTTGC.1         p1      11489         2617
# p1-AAACCTGGTACATGTC.1         p1      13608         2531
# p1-AAACCTGGTCGGCATC.1         p1      15664         3196
# p1-AAACCTGGTGCAGACA.1         p1      13066         2851
# p1-AAACCTGTCATTCACT.1         p1      22751         4221

scobj.raw <- scobj
scobj[["RNA"]] <- split(scobj[["RNA"]], f = scobj$orig.ident)

range(scobj$nFeature_RNA)
# [1]  132 5565
head(scobj@meta.data[,1:3])
# orig.ident nCount_RNA nFeature_RNA
# p1-AAACCTGAGAGCTATA.1         p1       4461         2095
# p1-AAACCTGAGTGTTTGC.1         p1       5375         2473
# p1-AAACCTGGTACATGTC.1         p1       6072         2397
# p1-AAACCTGGTCGGCATC.1         p1       7685         3047
# p1-AAACCTGGTGCAGACA.1         p1       6357         2708
# p1-AAACCTGTCATTCACT.1         p1      11800         4062

# sum(scobj$nFeature_RNA != scobj.raw$nFeature_RNA)
# # [1] 21516
# data.frame(
#   raw = scobj.raw$nFeature_RNA,
#   spl = scobj$nFeature_RNA
# )
# #                       raw  spl
# # p1-AAACCTGAGAGCTATA.1 2237 2095
# # p1-AAACCTGAGTGTTTGC.1 2617 2473
# # p1-AAACCTGGTACATGTC.1 2531 2397
# # p1-AAACCTGGTCGGCATC.1 3196 3047
# # p1-AAACCTGGTGCAGACA.1 2851 2708
# # p1-AAACCTGTCATTCACT.1 4221 4062
# # p1-AAACCTGTCTTGAGGT.1 2949 2808
# # p1-AAACGGGAGAGCCTAG.1 3547 3400
# # p1-AAACGGGGTAGGCATG.1 2535 2398
# # p1-AAACGGGGTCGACTAT.1 2035 1896

save(scobj, file = "scobj(GSE137829).Rdata")
# Nomalization & harmony & cluster----------------------------------------------
scobj.harmony <- scobj_list %>% 
  NormalizeData(normalization.method = "LogNormalize") %>% # vst.flavor = 'v2', verbose = FALSE
  FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>% 
  ScaleData %>% 
  RunPCA(npcs = 50) %>% 
  RunHarmony(
    group.by.vars = "orig.ident",
    reduction.use = "pca",
    reduction.save = "harmony") %>% 
  JoinLayers(assay = "RNA")

scobj.harmony.20 <- scobj.harmony %>%
  FindNeighbors(reduction = "harmony", dims = 1:20) %>%
  FindClusters(resolution = seq(0.1, 1, 0.1))
scobj.harmony.25 <- scobj.harmony %>%
  FindNeighbors(reduction = "harmony", dims = 1:25) %>%
  FindClusters(resolution = seq(0.1, 1, 0.1))
scobj.harmony.30 <- scobj.harmony %>%
  FindNeighbors(reduction = "harmony", dims = 1:30) %>%
  FindClusters(resolution = seq(0.1, 1, 0.1))

pdf("vst.2000.h clustree 20 25 30.pdf")
ElbowPlot(scobj.harmony, reduction = "pca", ndims = 50)
clustree(scobj.harmony.20, prefix = "RNA_snn_res.")
clustree(scobj.harmony.25, prefix = "RNA_snn_res.")
clustree(scobj.harmony.30, prefix = "RNA_snn_res.")
dev.off()

scobj.harmony <- scobj.harmony.30 %>% 
  RunUMAP(reduction = "harmony", dims = 1:30)
# Visualization-----------------------------------------------------------------
scobj.harmony <- RegroupIdents(scobj.harmony, metadata = "RNA_snn_res.0.2")
DimPlot(scobj.harmony, group.by = "orig.ident", reduction = "umap", label = T)
DimPlot(scobj.harmony, group.by = "RNA_snn_res.0.2", reduction = "umap", label = T)
# Cell Cycle--------------------------------------------------------------------
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

scobj.harmony <- CellCycleScoring(scobj.harmony, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

pdf("cell cycle.pdf")
RidgePlot(scobj.harmony, features = c("PCNA", "TOP2A", "MCM6", "MKI67"), ncol = 2)
scobj.harmony <- RunPCA(scobj.harmony, features = c(s.genes, g2m.genes), reduction.name = "cellcycle")
DimPlot(scobj.harmony, reduction = "cellcycle")
DimPlot(scobj.harmony, reduction = "umap")
dev.off()
# doublets----------------------------------------------------------------------
save(scobj.harmony, file = "scobj.harmony(GSE137829).Rdata")

if(TRUE){
  require(DoubletFinder)
  scobj.harmony.split <- SplitObject(scobj.harmony, split.by = "orig.ident") # into list
  
  for (i in 1:length(scobj.harmony.split)) {
    # pK Identification (ground-truth) -------------------------------------------
    sweep.list <- paramSweep(scobj.harmony.split[[i]], PCs = 1:30)
    sweep.stats <- summarizeSweep(sweep.list, GT = FALSE)
    bcmvn <- find.pK(sweep.stats)
    
    pK <- as.numeric(as.vector(bcmvn$pK[which.max(bcmvn$BCmetric)])) 
    ## Homotypic Doublet Proportion Estimate -------------------------------------
    homotypic.prop <- modelHomotypic(scobj.harmony.split[[i]]@meta.data$RNA_snn_res.0.2)  
    ## Assuming 7.5% doublet formation rate - tailor for your dataset
    nExp_poi <- round(0.05 * nrow(scobj.harmony.split[[i]]@meta.data))  ## Assuming 5% doublet formation rate - tailor for your dataset
    nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
    ## Run DoubletFinder with varying classification stringencies ----------------
    # scobj.harmony.split[[i]] <- doubletFinder(scobj.harmony.split[[i]], PCs = 1:30, pN = 0.25, pK = pK, nExp = nExp_poi, reuse.pANN = FALSE, sct = TRUE)
    
    # https://github.com/chris-mcginnis-ucsf/DoubletFinder/issues/228
    # https://github.com/chris-mcginnis-ucsf/DoubletFinder/issues/225#issuecomment-2786505997
    scobj.harmony.split[[i]] <- doubletFinder(
      scobj.harmony.split[[i]], PCs = 1:30, pN = 0.25, pK = pK, 
      nExp = nExp_poi.adj, sct = FALSE)
  }
  
  save(scobj.harmony.split, file = "scobj.harmony.split(GSE137829).Rdata")
  
  Singlet <- c(
    rownames(scobj.harmony.split[[1]]@meta.data) [scobj.harmony.split[[1]]@meta.data$DF.classifications_0.25 == "Singlet"],
    rownames(scobj.harmony.split[[2]]@meta.data) [scobj.harmony.split[[2]]@meta.data$DF.classifications_0.25 == "Singlet"],
    rownames(scobj.harmony.split[[3]]@meta.data) [scobj.harmony.split[[3]]@meta.data$DF.classifications_0.25 == "Singlet"],
    rownames(scobj.harmony.split[[4]]@meta.data) [scobj.harmony.split[[4]]@meta.data$DF.classifications_0.25 == "Singlet"],
    rownames(scobj.harmony.split[[5]]@meta.data) [scobj.harmony.split[[5]]@meta.data$DF.classifications_0.25 == "Singlet"],
    rownames(scobj.harmony.split[[6]]@meta.data) [scobj.harmony.split[[6]]@meta.data$DF.classifications_0.25 == "Singlet"])
  
  scobj.harmony@meta.data$id <- rownames(scobj.harmony@meta.data)
  
  # scobj.h.dblfinder <- subset(scobj.harmony, subset = id %in% Singlet)
  scobj.harmony@meta.data$dblfinder <- NA
  scobj.harmony@meta.data$dblfinder <- "doublet"
  scobj.harmony@meta.data$dblfinder[scobj.harmony@meta.data$id %in% Singlet] <- "singlet"
  
  dim(scobj.harmony)
  # [1] 15486 21516
  table(scobj.harmony@meta.data$dblfinder)
  # doublet singlet 
  # 784   20732
}
# Re-run========================================================================
scobj[['S.Score']] <- scobj.harmony$S.Score
scobj[['G2M.Score']] <- scobj.harmony$G2M.Score
scobj[['Phase']] <- scobj.harmony$Phase
scobj[['dblfinder']] <- scobj.harmony$dblfinder

save(scobj, file ="scobj_for_2nd(GSE137829).Rdata")

# 1000
if(FALSE){
  scobj.harmony.1000 <- scobj %>% 
    subset(subset = dblfinder == "singlet") %>%
    NormalizeData(normalization.method = "LogNormalize") %>% # vst.flavor = 'v2', verbose = FALSE
    FindVariableFeatures(selection.method = "vst", nfeatures = 1000) %>% 
    ScaleData(vars.to.regress = c("S.Score", "G2M.Score"), 
              features = rownames(scobj)) %>% 
    RunPCA(npcs = 50) %>% 
    RunHarmony(
      group.by.vars = "orig.ident",
      reduction.use = "pca",
      reduction.save = "harmony") %>% 
    JoinLayers(assay = "RNA")
  
  scobj.harmony.1000.20 <- scobj.harmony.1000 %>%
    FindNeighbors(reduction = "harmony", dims = 1:20) %>%
    FindClusters(resolution = seq(0.1, 1, 0.1))
  scobj.harmony.1000.25 <- scobj.harmony.1000 %>%
    FindNeighbors(reduction = "harmony", dims = 1:25) %>%
    FindClusters(resolution = seq(0.1, 1, 0.1))
  scobj.harmony.1000.30 <- scobj.harmony.1000 %>%
    FindNeighbors(reduction = "harmony", dims = 1:30) %>%
    FindClusters(resolution = seq(0.1, 1, 0.1))
  
  pdf("vst.1000.h_clustree 20 25 30(singlet_regcc).pdf")
  ElbowPlot(scobj.harmony.1000, reduction = "pca", ndims = 50)
  clustree(scobj.harmony.1000.20, prefix = "RNA_snn_res.")
  clustree(scobj.harmony.1000.25, prefix = "RNA_snn_res.")
  clustree(scobj.harmony.1000.30, prefix = "RNA_snn_res.")
  dev.off()
  
  save(scobj.harmony.1000, 
       scobj.harmony.1000.20, scobj.harmony.1000.25, scobj.harmony.1000.30,
       file = "scobj.harmony.1000(reg.cc).Rdata")
}
# 1500
if(FALSE){
  scobj.harmony.1500 <- scobj %>% 
    subset(subset = dblfinder == "singlet") %>%
    NormalizeData(normalization.method = "LogNormalize") %>% # vst.flavor = 'v2', verbose = FALSE
    FindVariableFeatures(selection.method = "vst", nfeatures = 1500) %>% 
    ScaleData(vars.to.regress = c("S.Score", "G2M.Score"), 
              features = rownames(scobj)) %>% 
    RunPCA(npcs = 50) %>% 
    RunHarmony(
      group.by.vars = "orig.ident",
      reduction.use = "pca",
      reduction.save = "harmony") %>% 
    JoinLayers(assay = "RNA")
  
  scobj.harmony.1500.20 <- scobj.harmony.1500 %>%
    FindNeighbors(reduction = "harmony", dims = 1:20) %>%
    FindClusters(resolution = seq(0.1, 1, 0.1))
  scobj.harmony.1500.25 <- scobj.harmony.1500 %>%
    FindNeighbors(reduction = "harmony", dims = 1:25) %>%
    FindClusters(resolution = seq(0.1, 1, 0.1))
  scobj.harmony.1500.30 <- scobj.harmony.1500 %>%
    FindNeighbors(reduction = "harmony", dims = 1:30) %>%
    FindClusters(resolution = seq(0.1, 1, 0.1))
  
  pdf("vst.1500.h_clustree 20 25 30(singlet_regcc).pdf")
  ElbowPlot(scobj.harmony.1500, reduction = "pca", ndims = 50)
  clustree(scobj.harmony.1500.20, prefix = "RNA_snn_res.")
  clustree(scobj.harmony.1500.25, prefix = "RNA_snn_res.")
  clustree(scobj.harmony.1500.30, prefix = "RNA_snn_res.")
  dev.off()
  
  save(scobj.harmony.1500, 
       scobj.harmony.1500.20, scobj.harmony.1500.25, scobj.harmony.1500.30,
       file = "scobj.harmony.1500(reg.cc).Rdata")
}
# 2000
if(FALSE){
  scobj.harmony.2000 <- scobj %>% 
    subset(subset = dblfinder == "singlet") %>%
    NormalizeData(normalization.method = "LogNormalize") %>% # vst.flavor = 'v2', verbose = FALSE
    FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>% 
    ScaleData(vars.to.regress = c("S.Score", "G2M.Score"), 
              features = rownames(scobj)) %>% 
    RunPCA(npcs = 50) %>% 
    RunHarmony(
      group.by.vars = "orig.ident",
      reduction.use = "pca",
      reduction.save = "harmony") %>% 
    JoinLayers(assay = "RNA")
  
  scobj.harmony.2000.20 <- scobj.harmony.2000 %>%
    FindNeighbors(reduction = "harmony", dims = 1:20) %>%
    FindClusters(resolution = seq(0.1, 1, 0.1))
  scobj.harmony.2000.25 <- scobj.harmony.2000 %>%
    FindNeighbors(reduction = "harmony", dims = 1:25) %>%
    FindClusters(resolution = seq(0.1, 1, 0.1))
  scobj.harmony.2000.30 <- scobj.harmony.2000 %>%
    FindNeighbors(reduction = "harmony", dims = 1:30) %>%
    FindClusters(resolution = seq(0.1, 1, 0.1))
  
  pdf("vst.2000.h_clustree 20 25 30(singlet_regcc).pdf")
  ElbowPlot(scobj.harmony.2000, reduction = "pca", ndims = 50)
  clustree(scobj.harmony.2000.20, prefix = "RNA_snn_res.")
  clustree(scobj.harmony.2000.25, prefix = "RNA_snn_res.")
  clustree(scobj.harmony.2000.30, prefix = "RNA_snn_res.")
  dev.off()
  
  save(scobj.harmony.2000, 
       scobj.harmony.2000.20, scobj.harmony.2000.25, scobj.harmony.2000.30,
       file = "scobj.harmony.2000(reg.cc).Rdata")
}

if(regression cell cycle){
  scobj.harmony <- scobj.harmony.1000.30 %>% 
    RunUMAP(reduction = "harmony", dims = 1:30)
  # Visualization=================================================================
  scobj.harmony <- RegroupIdents(scobj.harmony, metadata = "RNA_snn_res.0.4")
  # DimPlot(scobj.harmony, group.by = "orig.ident", reduction = "umap", label = T)
  # DimPlot(scobj.harmony, group.by = "RNA_snn_res.0.4", reduction = "umap", label = T)
  # Annotation====================================================================
  scobj.harmony <- RegroupIdents(scobj.harmony, metadata = "RNA_snn_res.0.4")
  
  epithelial <- c("EPCAM", "SFN", "KRT5", "KRT14","KLK3") #, "SPRR3"
  endothelial <- c("VWF", "PECAM1", "ENG", "CDH5") #, "CCL14"
  fibroblast <- c("FN1", "DCN", "COL1A1", "COL1A2") #, "COL3A1", "COL6A1"
  SMC <- c("TAGLN", "CNN1", "PRKG1", "FOXP2")
  pericyte <- c("RGS5", "MCAM", "ACTA2", "MYH11")
  T_cell <- c("CD2", "CD3D", "CD3E", "CD3G") # 0 4 10 14 
  B_cell <-	c("CD19", "CD79A", "MS4A1", "CD79B") # 12
  plasma <-	c("JCHAIN", "MZB1", "IGHG1", "SDC1") # 8 16 , "CD79A"
  # monocyte and macrophage
  myeloid <- c("CD68", "CD163", "LYZ", "CD14","FCGR3A","C1QA","C1QB","CST3") # , "IL3RA", "LAMP3", "CLEC4C", "GCA"
  Neutrophil <- c("CSF3R","CXCL8","G0S2","IFITM2") # "FCGR3B","FPR1","BASP1","CXCR1","CXCR2","S100A11"
  DC <- c('CCR7',	'CLEtmpA', 'CD1C',	'IRF7',	'LILRA4')
  Mast <-	c('TPSAB1','CPA3','HPGDS','KIT')# 'VWA5A','SLC18A2','HDC','CAPG','RGS13','IL1RL1','FOSB','GATA2'
  Neural <- c("PLP1","NRNX1","NRNX2","NRNX3")
  Proliferation <- c("TOP2A","MKI67","PCNA","BIRC5","DLGAP5")
  
  p <- DotPlot(scobj.harmony, features = c(
    epithelial, endothelial, fibroblast, SMC, pericyte,
    T_cell, B_cell,plasma, myeloid, Neutrophil, Mast, DC, Neural,
    Proliferation
  ) %>% unique, 
  group.by = "RNA_snn_res.0.4") + 
    scale_color_viridis() +
    labs(x = "", y = "") +
    coord_flip() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  ggsave(p, filename = "Dotplot marker(GSE137829).pdf", width=8, height=8, units="in")
    
  scobj.harmony$cell_type <- "Unknown"
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(0)] <- "Epithelial"
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(1)] <- "T cell"
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(2)] <- "Mesenchyme" # Fibroblast
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(3)] <- "Epithelial"
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(4)] <- "Myeloid"
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(5)] <- "Mesenchyme" # SMC
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(6)] <- "Endothelial"
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(7)] <- "Mast"
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(8)] <- "Epithelial"
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(9)] <- "Epithelial"
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(10)]<- "Proliferation"
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(11)]<- "Epithelial"
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(12)]<- "Proliferation"
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(13)]<- "Plasma"
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(14)]<- "B cell"
  scobj.harmony$cell_type[scobj.harmony$RNA_snn_res.0.4 %in% c(15)]<- "T cell"
  
  # save(scobj.harmony, file = "scobj.harmony(singlet-reg.cc-1000-30-res_0.4) anno.Rdata")
  load("scobj.harmony(singlet-reg.cc-2000-25) anno.Rdata")
  
  table(scobj.harmony$RNA_snn_res.0.4)
  # 0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15
  # 6642 2459 1952 1658 1347 1296 1041  974  761  650  482  387  373  302  292  116
  table(scobj.harmony$cell_type)
  # B cell   Endothelial    Epithelial          Mast    Mesenchyme       Myeloid 
  # 292          1041         10098           974          3248          1347 
  # Plasma Proliferation        T cell 
  # 302           855          2575 
  
  C9 <- FindMarkers(scobj.harmony, group.by = "RNA_snn_res.0.4", ident.1 = "9")  
  C10 <- FindMarkers(scobj.harmony, group.by = "RNA_snn_res.0.4", ident.1 = "10")  
  C12 <- FindMarkers(scobj.harmony, group.by = "RNA_snn_res.0.4", ident.1 = "12")  
  C15 <- FindMarkers(scobj.harmony, group.by = "RNA_snn_res.0.4", ident.1 = "15")
  
  tmp <- C15
  head(tmp[tmp$p_val_adj < 0.05 & tmp$avg_log2FC > 0 & tmp$pct.1 > 0.2,], 100)
  
  pdf("check-cluster9.pdf")
  VlnPlot(scobj.harmony, group.by = "RNA_snn_res.0.4", features = c("nFeature_RNA"), ncol = 1)
  VlnPlot(scobj.harmony, group.by = "RNA_snn_res.0.4", features = c("nCount_RNA"), ncol = 1)
  FeatureScatter(scobj.harmony, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
  dev.off()
  # Visualization ==============================================================
  p1 <- DimPlot(scobj.harmony, reduction = "umap", 
                group.by = "cell_type", 
                label = TRUE)
  p2 <- DimPlot(scobj.harmony, reduction = "umap", 
                group.by = "orig.ident", 
                label = TRUE)
  p_out1 <- p1 + p2
  ggsave(p_out1, filename = "UMAP(GSE137829).pdf", width=10, height=5, units="in")
  
  p_fea <- lapply(c(gene_symbols), 
                  function(x){FeaturePlot(scobj.harmony, features = x) + 
                      scale_colour_gradientn(
                        colours = colorRampPalette(
                          c('gray90','#FFCA62','#FFB336','#FF9700','#FF5A00','#F24410',
                            '#E52C22','#DD1D23','#C20030','#930039','#8C003A',
                            '#6F003D','#56033F'))(1000))}) %>% 
    patchwork::wrap_plots(ncol = 2, nrow = 2)
  p_vln <- VlnPlot(scobj.harmony, c(gene_symbols), group.by = "cell_type")
  
  ggsave(p1, filename = "UMAP(GSE137829).pdf", width=6, height=5, units="in")
  ggsave(p_fea, filename = "FeaturePlot(GSE137829).pdf", width=8, height=8, units="in")
  ggsave(p_vln, filename = "Vln_plot(GSE137829).pdf", width=9, height=6, units="in")
}






















