# NicheNet
# use data from GSE115469

# This package uses a count matrix and metadata generated by single-cell RNAseq 
# to evaluate putative cell-cell communication. It does this by utilizing a list 
# of interacting ligand and receptor pairs described in "A draft network of 
# ligand-receptor-mediated multicellular signaling in human" (Ramilowski JA et al, 
# Nature Communications 2015). Briefly, celltalker accomplishes this by building 
# lists of expressed ligands and receptors across cell types and sample groups, 
# then looks for cognate ligand/receptor expression within and between cell types. 
# Methods are provided for visualization of unique and differentially expressed 
# ligand and receptor interactions.

# Reference---------------------------------------------------------------------
# https://github.com/arc85/celltalker

# http://www.cellchat.org/
# https://github.com/jinworks/CellChat/
# https://www.jianshu.com/p/30c6e8a24415
# https://www.jianshu.com/p/03f9c2f3ee4f
# https://www.jianshu.com/p/38a9376f5286
# https://github.com/Teichlab/cellphonedb
# https://www.jianshu.com/p/dfe35a2a02d4
# Load pkgs---------------------------------------------------------------------
library(dplyr)
library(Seurat)
library(patchwork)
library(clustree) # use for determine the optimal resolution
library(ROGUE) # use for determine the optimal resolution
library(harmony)
library(stringr)
library(Matrix)
library(scDblFinder)
## CELL CHAT
# devtools::install_github("sqjin/CellChat")
library(CellChat)
# devtools::install_github("saeyslab/nichenetr")
library(nichenetr)
# devtools::install_github("Coolgenome/iTALK", build_vignettes = TRUE)
library(iTALK)
# BiocManager::install("SingleCellSignalR")
library(SingleCellSignalR)
# devtools::install_github("arc85/celltalker")
library(celltalker)
# devtools::install_github("mkarikom/RSoptSC")
library(RSoptSC)
# Load data---------------------------------------------------------------------
load("GSE115469-scobj.h.sc.annot.Rdata")

# InstallData("hcabm40k")   
# data(hcabm40k) 
# hcabm40k <- UpdateSeuratObject(hcabm40k) 

# Filter cell and assign cell types to the dataset------------------------------
# # NB: hca_bm_umap_cell_types has cell types and UMAP embeddings
# hca_bm <- hcabm40k[,rownames(hca_bm_umap_cell_types)]
# hca_bm[["cell_types"]] <- hca_bm_umap_cell_types$cell_types
# 
# # Process data
# hca_bm <- NormalizeData(hca_bm)
# 
# # Add UMAP coordinates
# hca_bm[["umap"]] <- CreateDimReducObject(embeddings=as.matrix(hca_bm_umap_cell_types[,1:2]),
#                                          key="UMAP_",assay="RNA")
# 
# # View cell types
# DimPlot(hca_bm,group.by="cell_types")

#- Run celltalker---------------------------------------------------------------
scobj_interactions <- celltalk(
  input_object = scobj.h.sc, # Seurat object to create ligand receptor matrices from
  metadata_grouping = "cell_type", # Seurat object to create ligand receptor matrices from
  ligand_receptor_pairs = ramilowski_pairs, # Data.frame of ligands, receptors and interactions in the format of ramilowski_pairs provided by this package. Defaults is "ramilowski_pairs".
  number_cells_required = 100,
  min_expression = 1000,
  max_expression = 20000,
  scramble_times = 10
  )

scobj_interactions
# A tibble: 3,960 × 9
interaction_pairs       interaction value scram_mean scram_sd  p_val cell_type1 cell_type2
<chr>                   <chr>       <dbl>      <dbl>    <dbl>  <dbl> <chr>      <chr>     
  1 Endothelia_Endothelia   A2M_LRP1    0.712      0.512    0.349 0.284  Endothelia Endothelia
2 Cholangiocyte_Endothel… A2M_LRP1    0.129      0.534    0.338 0.885  Cholangio… Endothelia
3 Mononuclear phagocytes… A2M_LRP1    0.245      0.469    0.306 0.768  Mononucle… Endothelia
4 Innate lymphoid cell_E… A2M_LRP1    0.101      0.519    0.309 0.911  Innate ly… Endothelia
5 Hepatocyte_Endothelia   A2M_LRP1    0.742      0.521    0.231 0.170  Hepatocyte Endothelia
6 B & Plasma_Endothelia   A2M_LRP1    0.196      0.446    0.320 0.782  B & Plasma Endothelia
7 Endothelia_Cholangiocy… A2M_LRP1    0.765      0.447    0.209 0.0643 Endothelia Cholangio…
8 Cholangiocyte_Cholangi… A2M_LRP1    0.183      0.465    0.247 0.873  Cholangio… Cholangio…
9 Mononuclear phagocytes… A2M_LRP1    0.298      0.484    0.312 0.724  Mononucle… Cholangio…
10 Innate lymphoid cell_C… A2M_LRP1    0.154      0.501    0.291 0.883  Innate ly… Cholangio…
# ℹ 3,950 more rows
# ℹ 1 more variable: interact_ratio <dbl>
# ℹ Use `print(n = ...)` to see more rows

## Identify top statistically significant interactions
top_stats <- scobj_interactions %>%
  mutate(fdr = p.adjust(p_val, method = "fdr")) %>%
  filter(fdr < 0.05) %>%
  group_by(cell_type1) %>%
  top_n(3, interact_ratio) %>%
  ungroup()

scobj_interactions %>%
  mutate(fdr = p.adjust(p_val, method = "fdr")) %>%
  filter(fdr < 0.05) %>%
  group_by(cell_type1) %>%
  summarise(n = n())
# A tibble: 6 × 2
cell_type1                 n
<chr>                  <int>
  1 B & Plasma                 5
2 Cholangiocyte              5
3 Endothelia                 7
4 Hepatocyte                36
5 Innate lymphoid cell      12
6 Mononuclear phagocytes    14

table(top_stats$cell_type1)
# B & Plasma          Cholangiocyte             Endothelia             Hepatocyte 
# 3                      3                      3                      3 
# Innate lymphoid cell Mononuclear phagocytes 
# 3                      3 

## Generate a circos plot
colors_use <- RColorBrewer::brewer.pal(n = length(unique(scobj.h.sc$cell_type)), "Set2")

circos_plot(
  ligand_receptor_frame = top_stats,
  cell_group_colors = colors_use,
  ligand_color = "blue",
  receptor_color = "red",
  cex_outer = 0.5,
  cex_inner = 0.4
  )













